docs(guide) add Prometheus Grafana integration guide (#353)
Kong/kubernetes-ingress-controller: docs/guides/prometheus-grafana.md

                      # to the Prometheus Server we installed previous
     isDefault: true  # The default data source is Prometheus

dashboardProviders:
  dashboardproviders.yaml:
    apiVersion: 1
    providers:
    - name: 'default' # Configure a dashboard provider file to
      orgId: 1        # put Kong dashboard into.
      folder: ''
      type: file
      disableDeletion: false
      editable: true
      options:
        path: /var/lib/grafana/dashboards/default
dashboards:
  default:
    kong-dash:
      gnetId: 7424  # Install the following Grafana dashboard in the
      revision: 5   # instance: https://grafana.com/dashboards/7424
      datasource: Prometheus
```

To install Grafana, execute the following:

```bash
helm install stable/grafana --name grafana --namespace monitoring --values http://bit.ly/2FuFVfV --version 1.22.1
```

## Install Kong Ingress Controller

We will use Kong's Helm chart to install Kong Ingress Controller
but you can also use plain manifests for this purpose.

```bash
helm install stable/kong --name kong --namespace kong --values https://bit.ly/2YDHyoh
```

### Enable Prometheus plugin in Kong

> We will enable the Promtheus plugin in Kong at the global level, meaning,

Reviewer: ```suggestion
We will enable the Promtheus plugin in Kong at the global level, meaning
```

each request that flows into the Kubernetes cluster gets tracked in Prometheus:

```bash
echo "apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  labels:
    global: \"true\"
  name: prometheus
plugin: prometheus
" | kubectl apply -f -
```

## Set Up Port Forwards

Now, we will gain access to the components we just deployed.
In a production environment, you would have a Kubernetes Service with

> external IP or load balancer, which would allow you to access

Reviewer: ```suggestion
an external IP or load balancer, which would allow you to access
```

> Prometheus, Grafana and Kong.

Reviewer: ```suggestion
Prometheus, Grafana, and Kong.
```

For demo purposes, we will set up port-forwarding using kubectl to get access.

> Please do not do this in production.

Reviewer: ```suggestion
It is not advisable to do this in production.
```

Open a new terminal and execute the following commands:

```bash
POD_NAME=$(kubectl get pods --namespace monitoring -l "app=prometheus,component=server" -o jsonpath="{.items[0].metadata.name}")
kubectl --namespace monitoring  port-forward $POD_NAME 9090 &

# You can access Prometheus in your browser at localhost:9090

POD_NAME=$(kubectl get pods --namespace monitoring -l "app=grafana" -o jsonpath="{.items[0].metadata.name}")
kubectl --namespace monitoring port-forward $POD_NAME 3000 &

# You can access Grafana in your browser at localhost:3000

POD_NAME=$(kubectl get pods --namespace kong -l "app=kong" -o jsonpath="{.items[0].metadata.name}")
kubectl --namespace kong port-forward $POD_NAME 8000 &

# Kong proxy port is now your localhost 8000 port
# We are using plain-text HTTP proxy for this purpose of
# demo.
# You can also use the LoadBalancer IP address and not set up this
# port-forwarding if you are running in a cloud environment.
```

## Access Grafana Dashboard

To access Grafana, you need to get the password for the admin user.

> Execute the following to read the password and take a note of it:

Reviewer: ```suggestion
Execute the following to read the password and take note of it:
```

```bash
kubectl get secret --namespace monitoring grafana -o jsonpath="{.data.admin-password}" | base64 --decode ; echo
```

Now, browse to [http://localhost:3000](http://localhost:3000) and
fill in username as “admin” and password as what you just noted above.
You should be logged in to Grafana and Kong’s Grafana Dashboard
should already be installed for you.

## Setup Services

We have all the components for monitoring installed,
we will now spin up some services for demo purposes and setup Ingress
routing for them.

### Install Services

> We will setup three services: billing, invoice, comments.

Reviewer: ```suggestion
We will set up three services: billing, invoice, and comments.
```

Execute the following to spin these services up:

```bash
kubectl apply -f https://gist.githubusercontent.com/hbagdi/2d8ef66fe22cb99e1514f410f992268d/raw/a03d789b70c46ccd0b99d9f1ed838dc21419fc33/multiple-services.yaml
```

### Install Ingress for the Services

Next, once the services are up and running, we will create Ingress
routing rules in Kubernetes.
This will configure Kong to proxy traffic destined for these services correctly.

Execute the following:

```bash
echo "apiVersion: configuration.konghq.com/v1
kind: KongIngress
metadata:
  name: strip-path
route:
  strip_path: true
" | kubectl apply -f -

echo "apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    configuration.konghq.com: strip-path
  name: sample-ingresses
spec:
  rules:
  - http:
     paths:
     - path: /billing
       backend:
         serviceName: billing
         servicePort: 80
     - path: /comments
       backend:
         serviceName: comments
         servicePort: 80
     - path: /invoice
       backend:
         serviceName: invoice
         servicePort: 80" | kubectl apply -f -
```

## Let’s Create Some Traffic

We’re done configuring our services and proxies.

> Time to see if our set up works or catches fire.

Reviewer: ```suggestion
Time to see if our setup works or catches fire.
```

Reviewer: @hbagdi is there something more subtle we can say than "catches fire?" 

I think we can just say,

```suggestion
Time to see if our setup works.
```

Reviewer: (going with the latter)

Execute the following in a new terminal:

```bash
while true;
do
  curl http://localhost:8000/billing/status/200
  curl http://localhost:8000/billing/status/501
  curl http://localhost:8000/invoice/status/201
  curl http://localhost:8000/invoice/status/404
  curl http://localhost:8000/comments/status/200
  curl http://localhost:8000/comments/status/200
  sleep 0.01
done
```

Since we have already enabled Prometheus plugin in Kong to
collect metrics for requests proxied via Kong,
we should see metrics coming through in the Grafana dashboard.

You should be able to see metrics related to the traffic flowing
through our services.

> Try tweaking the above script to send different trafficpatterns

Reviewer: ```suggestion
Try tweaking the above script to send different traffic patterns
```

and see how the metrics change.
The upstream services are httpbin instances, meaning you can use
a variety of endpoints to shape your traffic.

## Metrics collected

### Request Latencies of Various Services

![Request latencies](../images/request-latencies.png)

Kong collects latency data of how long your services take to respond to
requests. One can use this data to alert the on-call engineer if the latency

> goes beyond a certain threshold. For example, let’s say you’ve an SLA

Reviewer: ```suggestion
goes beyond a certain threshold. For example, let’s say you have an SLA
```

that your APIs will respond with latency of less than 20 millisecond
for 95% of the requests.
You could configure Prometheus to alert based on the following query:

```text
histogram_quantile(0.95, sum(rate(kong_latency_bucket{type="request"}[1m])) by (le,service)) > 20
```

The query calculates the 95th percentile of the the total request
latency (or duration) for all of your services and alerts you if it is more
than 20 milliseconds.
The “type” label in this query is “request”, which tracks the latency
added by Kong and the service.

> You can switch this to “upstream”, to track latency added by the service only.

Reviewer: ```suggestion
You can switch this to “upstream” to track latency added by the service only.
```

> Prometheus is really flexible and well documented, so we won’t go into

Reviewer: ```suggestion
Prometheus is highly flexible and well documented, so we won’t go into
```

details of setting up alerts here, but you’ll be able to find them
in the Prometheus documentation.

### Kong Proxy Latency

![Proxy latencies](../images/proxy-latencies.png)

> Kong also collects metrics about it’s performance.

Reviewer: ```suggestion
Kong also collects metrics about its performance.
```

The following query is similar to the previous one but gives
us insight into latency added by Kong:

```text
histogram_quantile(0.90, sum(rate(kong_latency_bucket{type="kong"}[1m])) by (le,service)) > 2
```

### Error Rates

![Error rates](../images/error-rates.png)

Another important metric to track is the rate of errors and requests
your services are serving.

> The timeseries kong_http_status collects HTTP status code metrics

Reviewer: ```suggestion
The time series `kong_http_status` collects HTTP status code metrics
```

for each service.

This metric can help you track the rate of errors for each of your service:

```text
sum(rate(kong_http_status{code=~"5[0-9]{2}"}[1m])) by (service)
```

You can also calculate the percentage of requests in any duration
that are errors. Try to come up with a query to derive that result.

Please note that all HTTP status codes are indexed, meaning you could use
the data to learn about your typical traffic pattern and identify problems.
For example, a sudden rise in 404 response codes could be indicative
of client codes requesting an endpoint that was removed in a recent deploy.

### Request Rate and Bandwidth

![Request rates](../images/request-rate.png)

> One can derive the total request rate for each of your service or

Reviewer: ```suggestion
One can derive the total request rate for each of your services or
```

> across your Kubernetes cluster using the kong_http_status timeseries.

Reviewer: ```suggestion
across your Kubernetes cluster using the `kong_http_status` time series.
```

![Bandwidth](../images/bandwidth.png)

Another metric that Kong keeps track of is the amount of

> network bandwidth (kong_bandwidth) being consumed.

Reviewer: ```suggestion
network bandwidth (`kong_bandwidth`) being consumed.
```

This gives you an estimate of how request/response sizes

> co-relate with other behaviours in your infrastructure.

Reviewer: ```suggestion
correlate with other behaviors in your infrastructure.
```
